### ISSUE TITLE
#[derive] below #[salsa::tracked] is not functional?

### ISSUE THREAD

**Issue by NickNick** [salsa-rs/salsa#469]
Code like

```rust
#[salsa::tracked]
#[derive(what ever you write here is fine)]
struct Foo;
```

... will compile. It seems that derive line is discarded? I didn't notice until I added a derive I actually needed.

**Comment by xiyuzhai** [comment]
Strange. It never happens to me.

**Comment by Y-Nak** [comment]
> It seems that derive line is discarded?

Yes, `#[salsa::tracked]`/`#[salsa::input]`/`#[salsa::internred]` attribute macros intentionally drop `derive` macros, please refer to [here](https://github.com/salsa-rs/salsa/blob/f1d318a2795e87081e316a5f31ef02f0cd6a1c83/components/salsa-2022-macros/src/salsa_struct.rs#L166-L183) for more context.

**Comment by xiyuzhai** [comment]
That's intentional. The struct becomes just a thin wrapper of an id.

**Comment by xiyuzhai** [comment]
You can use rust analyzer expand macro command to see what's going on. Because the fields would require a database reference to access, the derive attr for Debug Clone Copy wouldn't make any sense. And by the way, the resulting type implements quite a few trait in its own way, Debug Clone Copy Hash

**Comment by NickNick** [comment]
Thanks for the responses.

However intentional it may be, it was surprising to me that it was just discarded.

How about an error for this, indicating that this isn't supported? Not sure if that is possible, but that would remove the surprise for me.

It is some time ago now, but if I recall correctly it did work when I flipped the lines:


```
#[derive(...)]
#[salsa(...)]
```

What I mean with that it did work is that that derive seemed effective, but I may not recall this correctly (and don't have the code anymore). The previously mentioned error could mention this as fix/workaround?

Thanks again!

**Comment by xiyuzhai** [comment]
I would imagine a fix would be quite simple. Just run through attrs in the procedural macro.
